recipes/src/recipes/RecipeForm.test.js

```diff
@@ -0,0 +1,153 @@
+    const descriptionInput = screen.getByDisplayValue(recipeUpdating.description);
+    expect(descriptionInputByLabel).toEqual(descriptionInput);
+
+    const ingredientsSelectBox = screen.getByTestId('ingredients-select-box');
+    recipeUpdating.ingredients.map(({
+      name
+    }) => expect(ingredientsSelectBox).toContainElement(screen.getByText(name)) );
+
+    expect(mockSetUpdating).not.toHaveBeenCalled();
+    expect(mockAddOrUpdateRecipe).not.toHaveBeenCalled();
+  });
+
+  test('it should display an empty form when "updating" does not match the id of any recipe', () => {
+    const updating = true;
+
+    const mockSetUpdating = jest.fn();
+    const mockAddOrUpdateRecipe = jest.fn();
+
+    const component = render(<RecipeForm recipes={recipes}
+                                         updating={updating}
+                                         setUpdating={mockSetUpdating}
+                                         addOrUpdateRecipe={mockAddOrUpdateRecipe}/>);
+
+    const nameInput = screen.getByLabelText('name', {exact: false});
+    expect(nameInput).not.toBeNull();
+    expect(nameInput).toHaveValue('');
+
+    const descriptionInput = screen.getByLabelText('description', {exact: false});
+    expect(descriptionInput).not.toBeNull();
+    expect(descriptionInput).toHaveValue('');
+
+    const ingredientsSelectBox = screen.getByTestId('ingredients-select-box');
+    const emptyIngredientsSelect = screen.getByText(INGREDIENTS_PLACEHOLDER , {
+      exact: false
+    });
+    expect(ingredientsSelectBox).toContainElement(emptyIngredientsSelect);
+
+    expect(mockSetUpdating).not.toHaveBeenCalled();
+    expect(mockAddOrUpdateRecipe).not.toHaveBeenCalled();
+  });
+
+  // test('it should call setUpdating(false) when the cancel button is clicked', async () => {
+  // });
+
+  test('it should pass the form details to addOrUpdateRecipe() when the submit button is clicked', async () => {
+    const updating = true;
+
+    const mockSetUpdating = jest.fn();
+    const mockAddOrUpdateRecipe = jest.fn();
+
+    const component = render(<RecipeForm recipes={recipes}
+                                         updating={updating}
+                                         setUpdating={mockSetUpdating}
+                                         addOrUpdateRecipe={mockAddOrUpdateRecipe}/>);
+
+    const name = 'Shepherds Pie';
+    const description = 'Layer mashed potatoes over mince in a casserole dish and bake';
+    const ingredients = [
+      'Lamb mince',
+      'Potatoes'
+    ];
+
+    await fillOutForm(screen, {
+      name,
+      description,
+      ingredients,
+    });
+
+    const button = screen.getByText('Submit');
+    expect(button).not.toBeNull();
+    fireEvent.click(button);
+
+    // expect(mockSetUpdating).toHaveBeenCalledWith(false);
+    //
+    // const expectedRecipe = {
+    //   name,
+    //   description,
+    //   ingredients,
+    // }
+    // expect(mockAddOrUpdateRecipe).toHaveBeenCalledWith(expectedRecipe);
```

Comment:
is this the test that is not working?

When testing code that calls async methods you have to wait for the async method to complete, otherwise the test executes before the async method and it fails:
```
await waitFor(() =>
      expect(mockAddOrUpdateRecipe).toHaveBeenCalledWith(expectedRecipe);
    )
```

Reply:
Ah - no, sorry, I seem to have forgotten to push some changes.
These tests all work now, but I suspect they were the ones that needed fixing 😄 .

The skipped test is in `RecipeTestComponent.test.js` - when clicking the Delete button, if I wait for the RecipeForm to be removed with `await waitForElementToBeRemoved(() => button);` it will timeout, but without that wait it still finds the components of the form. I tried returning a promise from the mock `deleteRecipeFromDb` and waiting for that, but I couldn't get it to work